### PR TITLE
Test: support push down filter in ut

### PR DESCRIPTION
**Pull request by ywqzzy**
### What problem does this PR solve?

Issue Number: ref #4609 

Problem Summary:
We don't push down filter to storage layer in current ut framework.
### What is changed and how it works?
In `getStreamFromDeltaMerge`, mock `DAGQueryInfo` with `PushDownFilter` from `tipb:Selection`
### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: dbms/src/Debug/MockStorage.cpp [hunk 1]
```diff
@@ -119,10 +124,11 @@ void MockStorage::addTableDataForDeltaMerge(Context & context, const String & na
     }
 }
 
-BlockInputStreamPtr MockStorage::getStreamFromDeltaMerge(Context & context, Int64 id)
+BlockInputStreamPtr MockStorage::getStreamFromDeltaMerge(Context & context, Int64 table_id, const PushDownFilter * push_down_filter)
```
Comment (Lloyd-Pottiger) [inline dbms/src/Debug/MockStorage.cpp:127]: why not use `shared_ptr`

File: dbms/src/Debug/MockStorage.cpp [hunk 2]
```diff
@@ -134,10 +140,29 @@ BlockInputStreamPtr MockStorage::getStreamFromDeltaMerge(Context & context, Int6
...
+        auto analyzer = std::make_unique<DAGExpressionAnalyzer>(names_and_types_map_for_delta_merge[table_id], context);
+        query_info.dag_query = std::make_unique<DAGQueryInfo>(
+            push_down_filter->conditions,
+            analyzer->getPreparedSets(),
+            analyzer->getCurrentInputColumns(),
+            context.getTimezoneInfo());
+        auto [before_where, filter_column_name, project_after_where] = ::DB::buildPushDownFilter(*push_down_filter, *analyzer);
+        BlockInputStreams ins = storage->read(column_names, query_info, context, stage, 8192, 1); // TODO: Support config max_block_size and num_streams
+        BlockInputStreamPtr in = ins[0];
+        in = std::make_shared<FilterBlockInputStream>(in, before_where, filter_column_name, "test");
```
Comment (Lloyd-Pottiger) [inline dbms/src/Debug/MockStorage.cpp:154]: add a assert for the size of `ins`

File: dbms/src/Flash/tests/gtest_filter_executor.cpp [hunk 3]
```diff
@@ -251,6 +252,38 @@ try
 }
 CATCH
 
+TEST_F(FilterExecutorTestRunner, PushDownFilter)
```
Comment (Lloyd-Pottiger) [inline dbms/src/Flash/tests/gtest_filter_executor.cpp:255]: maybe need more tests

File: dbms/src/Debug/MockStorage.cpp (hunk 2 above)
Comment (ywqzzy) [reply dbms/src/Debug/MockStorage.cpp:154]: The size will be changed by `num_streams`, so I didn't check it.
In future, I will pass `num_streams` into `storage->read()`

File: dbms/src/Flash/tests/gtest_filter_executor.cpp (hunk 3 above)
Comment (ywqzzy) [reply dbms/src/Flash/tests/gtest_filter_executor.cpp:255]: > maybe need more tests

Ok, I will test more filters

File: dbms/src/Debug/MockStorage.cpp (hunk 1 above)
Comment (ywqzzy) [reply dbms/src/Debug/MockStorage.cpp:127]: > why not use `shared_ptr`

The `filter` will be used only once

File: dbms/src/Flash/Planner/PhysicalPlan.cpp [hunk 4]
```diff
@@ -40,13 +41,21 @@ namespace DB
...
 {
-bool pushDownSelection(const PhysicalPlanNodePtr & plan, const String & executor_id, const tipb::Selection & selection)
+bool pushDownSelection(Context & context, const PhysicalPlanNodePtr & plan, const String & executor_id, const tipb::Selection & selection)
 {
     if (plan->tp() == PlanType::TableScan)
     {
         auto physical_table_scan = std::static_pointer_cast<PhysicalTableScan>(plan);
         return physical_table_scan->pushDownFilter(executor_id, selection);
     }
+    if (plan->tp() == PlanType::MockTableScan && context.isExecutorTest())
```
Comment (SeaRise) [inline dbms/src/Flash/Planner/PhysicalPlan.cpp:51]:
```suggestion
    if (unlikely(plan->tp() == PlanType::MockTableScan && context.isExecutorTest()))
```

**Review by SeaRise — COMMENTED** [review]
Others LGTM

**Review by SeaRise — APPROVED** [review]
lgtm

**Review by Lloyd-Pottiger — APPROVED** [review]

**Pull request merged**